common/beerocks/bcl/source/beerocks_socket_event_loop.cpp

```diff
@@ -0,0 +1,359 @@
+/* SPDX-License-Identifier: BSD-2-Clause-Patent
+ *
+ * SPDX-FileCopyrightText: 2016-2020 the prplMesh contributors (see AUTHORS.md)
+ *
+ * This code is subject to the terms of the BSD+Patent license.
+ * See LICENSE file for more details.
+ */
+
+#include <bcl/beerocks_socket_event_loop.h>
+
+#include <sys/epoll.h>
+#include <sys/timerfd.h>
+#include <unistd.h>
+
+#include <easylogging++.h>
+
+namespace beerocks {
+
+//////////////////////////////////////////////////////////////////////////////
+////////////////////////// Local module definitions //////////////////////////
+//////////////////////////////////////////////////////////////////////////////
+
+// Maximal number of events to process in a single epoll_wait call
+static constexpr int MAX_POLL_EVENTS = 17;
+
+//////////////////////////////////////////////////////////////////////////////
+/////////////////////////////// Implementation ///////////////////////////////
+//////////////////////////////////////////////////////////////////////////////
+
+SocketEventLoop::SocketEventLoop(std::chrono::milliseconds timeout) : m_timeout(timeout)
+{
+    m_epoll_fd = epoll_create1(0);
+    LOG_IF(m_epoll_fd == -1, FATAL) << "Failed creating epoll: " << strerror(errno);
+}
+
+SocketEventLoop::~SocketEventLoop()
+{
+    // Delete all the sockets in the poll
+    LOG(DEBUG) << "Removing " << m_fd_to_event_data.size()
+               << " FDs from the event loop: " << m_fd_to_event_data;
+
+    while (!m_fd_to_event_data.empty()) {
+        auto event_data = m_fd_to_event_data.begin()->second;
+        del_event(event_data->socket);
+    }
+
+    // Close the poll fd
+    LOG_IF(close(m_epoll_fd) == -1, ERROR)
+        << "Failed closing epoll file descriptor: " << strerror(errno);
+}
+
+bool SocketEventLoop::add_event(SocketEventLoop::EventType socket,
+                                const EventLoop::EventHandlers &handlers,
+                                std::chrono::milliseconds timeout)
+{
+    if (!socket) {
+        LOG(ERROR) << "Invalid socket pointer!";
+        return false;
+    }
+
+    // Make sure that the FD is not already part of the poll
+    if (m_fd_to_event_data.find(socket->getSocketFd()) != m_fd_to_event_data.end()) {
+        LOG(WARNING) << "Requested to add FD (" << socket->getSocketFd()
+                     << ") to the poll, but it's already there...";
+
+        return false;
+    }
+
+    // Build EventData structure
+    auto event_data      = std::make_shared<EventData>();
+    event_data->socket   = socket;
+    event_data->handlers = handlers;
+
+    // If timeout is set and a handler is defined, initialize a timerfd
+    if (timeout > std::chrono::milliseconds::zero()) {
+        if (handlers.on_timeout) {
+
+            // TODO: timerfd may not be supported on all kernels.
+            //       Implement a check at creation, and in case it's not supported
+            //       base the timeouts mechanism on the main timeout of the loop.
```

Comment:
Assuming that some systems won't support a more fine-grained timeouts does mean we shouldn't use them at all :)
The idea is that systems where timerfd is supported will be able to benefit from high resolution timeouts, while where it's not supported, the timeout will be based on the timeout of the event loop.

I do believe that it will be supported on most platforms, and that's why I didn't bother (yet) to implement the alternative.